File: apps/client/src/features/course/courses.ts
import { api, APIError } from "../../api";
import type {
  CourseConversationState,
  CourseCover,
  StoredCourse,
  CourseMaterial,
  StoredCourseConversation,
  OutlineClassification,
  OutlineReorganization,
} from "../../domain/learning";
import {
  ObjectStorageError,
  readObjectText,
  uploadObject,
} from "../../transport/object-storage";

export const emptyCourseState = (): CourseConversationState => ({
  messages: [],
  pages: [],
  presentations: [],
  currentPresentationId: "",
});

export async function listCourses() {
  return (await api<{ courses: StoredCourse[] }>("/courses")).courses;
}

export async function createCourse(
  title: string,
  topic: string,
  cover: CourseCover,
) {
  return (
    await api<{ course: StoredCourse }>("/courses", "POST", {
      title,
      topic,
      cover,
    })
  ).course;
}

export async function updateCourse(
  id: string,
  changes: { title?: string; topic?: string },
) {
  return (
    await api<{ course: StoredCourse }>(`/courses/${id}`, "PATCH", changes)
  ).course;
}

export async function deleteCourse(id: string) {
  await api(`/courses/${id}`, "DELETE");
}

export async function saveCourseConversation(
  course: StoredCourse,
  state: CourseConversationState,
) {
  await api(`/courses/${course.id}/conversation`, "PUT", {
    conversationId: course.conversationId,
    state,
  });
}

export async function createCourseConversation(
  courseId: string,
  sectionId: string,
  title: string,
) {
  return (
    await api<{ conversation: StoredCourseConversation }>(
      `/courses/${courseId}/sections/${sectionId}/conversations`,
      "POST",
      { title },
    )
  ).conversation;
}

export async function deleteCourseConversation(
  courseId: string,
  sectionId: string,
  conversationId: string,
) {
  return (
    await api<{ course: StoredCourse }>(
      `/courses/${courseId}/sections/${sectionId}/conversations/${conversationId}`,
      "DELETE",
    )
  ).course;
}

export async function replaceCourseOutline(
  courseId: string,
  sections: Array<{
    id?: string;
    title: string;
    objective: string;
    status?: "planned" | "active" | "complete" | "archived";
  }>,
) {
  return await api<
    | { course: StoredCourse }
    | { reorganization: OutlineReorganization }
  >(`/courses/${courseId}/outline`, "PUT", { sections });
}

export async function getCourseOutlineReorganization(courseId: string) {
  try {
    return (
      await api<{ reorganization: OutlineReorganization }>(
        `/courses/${courseId}/outline-reorganization`,
      )
    ).reorganization;
  } catch (error) {
    if (error instanceof APIError && error.status === 404) return null;
    throw error;
  }
}

export async function assignCourseOutlineConversation(
  courseId: string,
  reorganizationId: string,
  conversation: StoredCourseConversation,
  classification: OutlineClassification,
) {
  return await api<
    | { course: StoredCourse }
    | { reorganization: OutlineReorganization }
  >(
    `/courses/${courseId}/outline-reorganizations/${reorganizationId}/assignments/${conversation.id}`,
    "PUT",
    {
      ...classification,
      conversationUpdatedAt: conversation.updatedAt,
    },
  );
}

export type CourseOutlineClassifier = (
  reorganization: OutlineReorganization,
  conversation: StoredCourseConversation,
) => Promise<OutlineClassification>;

async function finishCourseOutlineReorganization(
  courseId: string,
  initial: OutlineReorganization,
  classify: CourseOutlineClassifier,
) {
  let reorganization = initial;
  classificationLoop: while (reorganization.pendingCount > 0) {
    const batch = reorganization.pending.slice(0, 3);
    if (!batch.length) throw new Error("课程大纲分类队列暂时不可用");
    const classifications = await Promise.all(
      batch.map(async (conversation) => ({
        conversation,
        classification: await classify(reorganization, conversation),
      })),
    );
    for (const { conversation, classification } of classifications) {
      let result;
      try {
        result = await assignCourseOutlineConversation(
          courseId,
          reorganization.id,
          conversation,
          classification,
        );
      } catch (error) {
        if (!(error instanceof APIError) || error.status !== 409) throw error;
        const refreshed = await getCourseOutlineReorganization(courseId);
        if (!refreshed) throw error;
        reorganization = refreshed;
        continue classificationLoop;
      }
      if ("course" in result) return result.course;
      reorganization = result.reorganization;
    }
  }
  throw new Error("课程大纲调整没有完成");
}

export async function reorganizeCourseOutline(
  courseId: string,
  sections: Parameters<typeof replaceCourseOutline>[1],
  classify: CourseOutlineClassifier,
) {
  const result = await replaceCourseOutline(courseId, sections);
  if ("course" in result) return result.course;
  return finishCourseOutlineReorganization(
    courseId,
    result.reorganization,
    classify,
  );
}

export async function resumeCourseOutlineReorganization(
  courseId: string,
  classify: CourseOutlineClassifier,
) {
  const reorganization = await getCourseOutlineReorganization(courseId);
  if (!reorganization) return null;
  return finishCourseOutlineReorganization(
    courseId,
    reorganization,
    classify,
  );
}

export async function listCourseMaterials(courseId: string) {
  return (
    await api<{ materials: CourseMaterial[] }>(`/courses/${courseId}/materials`)
  ).materials;
}

export async function uploadCourseMaterial(courseId: string, file: File) {
  const { upload } = await api<{
    upload: {
      id: string;
      url: string;
      headers: Record<string, string>;
      expiresAt: string;
    };
  }>(`/courses/${courseId}/material-uploads`, "POST", {
    name: file.name,
    sizeBytes: file.size,
  });
  try {
    await uploadObject(upload, file);
  } catch (error) {
    if (error instanceof ObjectStorageError) {
      throw new Error(
        error.kind === "network"
          ? "课程材料传输失败，请检查网络后重试"
          : "课程材料传输失败，请重新上传",
      );
    }
    throw error;
  }
  return (
    await api<{ material: CourseMaterial }>(
      `/courses/${courseId}/material-uploads/${upload.id}/complete`,
      "POST",
    )
  ).material;
}

export async function getCourseMaterial(courseId: string, materialId: string) {
  const download = await api<{
    material: CourseMaterial;
    url: string;
    headers?: Record<string, string>;
  }>(
    `/courses/${courseId}/materials/${materialId}/download`,
  );
  const content = await readObjectText(download);
  return { material: download.material, content };
}

export async function deleteCourseMaterial(
  courseId: string,
  materialId: string,
) {
  await api(`/courses/${courseId}/materials/${materialId}`, "DELETE");
}

File: apps/server/cmd/api/courses.go
package main

import (
	"encoding/json"
	"net/http"
	"time"

	appservice "github.com/LanternCX/zhiya/apps/server/internal/application"
	"github.com/LanternCX/zhiya/apps/server/internal/domain"
)

func (a *application) listCourses(w http.ResponseWriter, r *http.Request) {
	courses, err := a.courseService().List(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"))
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"courses": courses})
}

func (a *application) createCourse(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Title string `json:"title"`
		Topic string `json:"topic"`
		Cover struct {
			Motif   string `json:"motif"`
			Palette string `json:"palette"`
			Label   string `json:"label"`
		} `json:"cover"`
	}
	if err := a.readJSON(w, r, &input); err != nil {
		a.respondError(w, err)
		return
	}
	cover := domain.CourseCover{Motif: input.Cover.Motif, Palette: input.Cover.Palette, Label: input.Cover.Label}
	course, err := a.courseService().Create(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), input.Title, input.Topic, cover)
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, map[string]any{"course": course})
}

func (a *application) getCourse(w http.ResponseWriter, r *http.Request) {
	course, err := a.courseService().Get(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"))
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"course": course})
}

func (a *application) updateCourse(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Title *string `json:"title"`
		Topic *string `json:"topic"`
	}
	if err := a.readJSON(w, r, &input); err != nil {
		a.respondError(w, err)
		return
	}
	course, err := a.courseService().Update(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), input.Title, input.Topic)
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"course": course})
}

func (a *application) saveCourseConversation(w http.ResponseWriter, r *http.Request) {
	var input struct {
		ConversationID string          `json:"conversationId"`
		State          json.RawMessage `json:"state"`
	}
	if err := a.readJSON(w, r, &input); err != nil || input.ConversationID == "" || len(input.State) == 0 || !json.Valid(input.State) {
		if err == nil {
			err = bad("课程对话无效")
		}
		a.respondError(w, err)
		return
	}
	err := a.courseService().SaveConversation(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), input.ConversationID, input.State)
	a.respondOK(w, err)
}

func (a *application) replaceCourseOutline(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Sections []struct {
			ID        string `json:"id"`
			Title     string `json:"title"`
			Objective string `json:"objective"`
			Status    string `json:"status"`
		} `json:"sections"`
	}
	if err := a.readJSON(w, r, &input); err != nil {
		a.respondError(w, err)
		return
	}
	outline := make([]domain.OutlineSection, len(input.Sections))
	for index, section := range input.Sections {
		outline[index] = domain.OutlineSection{ID: section.ID, Title: section.Title, Objective: section.Objective, Status: section.Status}
	}
	result, err := a.courseService().ReplaceOutline(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), outline)
	if err != nil {
		a.respondError(w, err)
		return
	}
	if result.Reorganization != nil {
		writeJSON(w, http.StatusAccepted, map[string]any{"reorganization": result.Reorganization})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"course": result.Course})
}

func (a *application) getCourseOutlineReorganization(w http.ResponseWriter, r *http.Request) {
	reorganization, err := a.courseService().GetOutlineReorganization(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"))
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"reorganization": reorganization})
}

func (a *application) assignCourseOutlineConversation(w http.ResponseWriter, r *http.Request) {
	var input struct {
		SectionID             string `json:"sectionId"`
		ConversationUpdatedAt string `json:"conversationUpdatedAt"`
		Reason                string `json:"reason"`
		NewSection            *struct {
			Title     string `json:"title"`
			Objective string `json:"objective"`
		} `json:"newSection"`
	}
	if err := a.readJSON(w, r, &input); err != nil {
		a.respondError(w, err)
		return
	}
	updatedAt, err := time.Parse(time.RFC3339Nano, input.ConversationUpdatedAt)
	if err != nil {
		a.respondError(w, bad("课程对话分类无效"))
		return
	}
	var newSection *domain.OutlineSection
	if input.NewSection != nil {
		newSection = &domain.OutlineSection{Title: input.NewSection.Title, Objective: input.NewSection.Objective}
	}
	result, err := a.courseService().AssignOutlineConversation(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), r.PathValue("reorganizationId"), r.PathValue("conversationId"), input.SectionID, input.Reason, updatedAt, newSection)
	if err != nil {
		a.respondError(w, err)
		return
	}
	if result.Reorganization != nil {
		writeJSON(w, http.StatusAccepted, map[string]any{"reorganization": result.Reorganization})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"course": result.Course})
}

func (a *application) createCourseConversation(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Title string `json:"title"`
	}
	if err := a.readJSON(w, r, &input); err != nil {
		a.respondError(w, err)
		return
	}
	conversation, err := a.courseService().CreateConversation(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), r.PathValue("sectionId"), input.Title)
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, map[string]any{"conversation": conversation})
}

func (a *application) deleteCourseConversation(w http.ResponseWriter, r *http.Request) {
	course, err := a.courseService().DeleteConversation(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), r.PathValue("sectionId"), r.PathValue("conversationId"))
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"course": course})
}

func (a *application) listCourseMaterials(w http.ResponseWriter, r *http.Request) {
	materials, err := a.courseService().ListMaterials(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"))
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"materials": materials})
}

func (a *application) startCourseMaterialUpload(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Name      string `json:"name"`
		SizeBytes int64  `json:"sizeBytes"`
	}
	if err := a.readJSON(w, r, &input); err != nil {
		a.respondError(w, err)
		return
	}
	upload, err := a.courseService().StartUpload(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), input.Name, input.SizeBytes, a.courseWarning(w))
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, map[string]any{"upload": map[string]any{"id": upload.Record.ID, "url": upload.Request.URL, "headers": upload.Request.Headers, "expiresAt": upload.Record.ExpiresAt}})
}

func (a *application) completeCourseMaterialUpload(w http.ResponseWriter, r *http.Request) {
	material, err := a.courseService().CompleteUpload(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), r.PathValue("uploadId"), a.courseWarning(w))
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, map[string]any{"material": material})
}

func (a *application) courseWarning(w http.ResponseWriter) appservice.Warn {
	return func(message string, err error, values ...any) { a.warnRequest(w, message, err, values...) }
}

func (a *application) downloadCourseMaterial(w http.ResponseWriter, r *http.Request) {
	material, request, err := a.courseService().Download(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), r.PathValue("materialId"))
	if err != nil {
		a.respondError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"material": material, "url": request.URL, "headers": request.Headers})
}

func (a *application) deleteCourseMaterial(w http.ResponseWriter, r *http.Request) {
	err := a.courseService().DeleteMaterial(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"), r.PathValue("materialId"))
	a.respondOK(w, err)
}

func (a *application) deleteCourse(w http.ResponseWriter, r *http.Request) {
	err := a.courseService().Delete(r.Context(), sessionToken(r), r.Header.Get("X-Zhiya-User"), r.PathValue("id"))
	a.respondOK(w, err)
}
